### PR TITLE
*: update privilege Checker interface and use it

### PR DESCRIPTION
**Pull request by tiancaiamao**
* modify the privilege.Checker interface
* use new implementation to check privilege
* add command line flag to enable/disable privilege feature

### CODE REVIEW

**Comment by zimulala** [comment]
Fix CI about `should omit type bool from declaration of var Enable; it will be inferred from the right-hand side`

File: session.go [hunk 1]
```diff
@@ -715,8 +715,14 @@ func CreateSession(store kv.Storage) (Session, error) {
 		return nil, errors.Trace(err)
 	}
 
-	// TODO: Add auth here
-	privChecker := &privileges.UserPrivileges{}
+	// Add auth here
```
Comment (zimulala) [inline session.go]: Add `.` at the end of a line.

File: plan/optimizer.go [hunk 2]
```diff
@@ -39,17 +40,37 @@ func Optimize(ctx context.Context, node ast.Node, is infoschema.InfoSchema) (Pla
...
-		allocator: allocator}
+		allocator: allocator,
+		visitInfo: make([]visitInfo, 0, 3),
+	}
 	p := builder.build(node)
 	if builder.err != nil {
 		return nil, errors.Trace(builder.err)
 	}
+
+	// Maybe it's better put this in Preprocess, but check privilege need table
```
Comment (zimulala) [inline plan/optimizer.go]: s/put/to put

File: plan/planbuilder.go [hunk 3]
```diff
@@ -69,6 +76,8 @@ type planBuilder struct {
 	inUpdateStmt bool
 	// colMapper stores the column that must be pre-resolved.
 	colMapper map[*ast.ColumnNameExpr]int
+	// collect the visit information for privilege check
```
Comment (zimulala) [inline plan/planbuilder.go]: Complete the comment.
s/collect/Collect, add `.`

File: privilege/privileges/privileges.go [hunk 4]
```diff
@@ -167,8 +171,38 @@ func (ps *userPrivileges) ShowGrants() []string {
...
+	*Handle
+}
+
+// RequestVerification implements the Checker interface.
+func (p *UserPrivileges) RequestVerification(db, table, column string, priv mysql.PrivilegeType) bool {
+	if !Enable {
+		return true
+	}
+
+	if p.User == "" {
```
Comment (zimulala) [inline privilege/privileges/privileges.go]: s/p.User == ""/len(p.User) == 0

File: plan/optimizer.go [hunk 5]
```diff
@@ -39,17 +40,37 @@ func Optimize(ctx context.Context, node ast.Node, is infoschema.InfoSchema) (Pla
 		ctx:       ctx,
 		is:        is,
 		colMapper: make(map[*ast.ColumnNameExpr]int),
-		allocator: allocator}
+		allocator: allocator,
+		visitInfo: make([]visitInfo, 0, 3),
```
Comment (zimulala) [inline plan/optimizer.go]: Why is it 3?

File: privilege/privileges/privileges.go (hunk 4 above)
Comment (tiancaiamao) [inline privilege/privileges/privileges.go]: what's the difference?

Comment (zimulala) [inline privilege/privileges/privileges.go]: This method of `len(p.User) == 0` is more commonly used. Anyway, it's up to you.

**Comment by zimulala** [comment]
LGTM

File: privilege/privileges/cache.go [hunk 6]
```diff
@@ -393,18 +393,26 @@ func (p *MySQLPrivilege) RequestVerification(user, host, db, table, column strin
...
+		ctx: ctx,
+	}
+}
+
 // Get the MySQLPrivilege for read.
 func (h *Handle) Get() *MySQLPrivilege {
 	return h.priv.Load().(*MySQLPrivilege)
 }
 
 // Update the MySQLPrivilege.
```
Comment (shenli) [inline privilege/privileges/cache.go]: The comment is not clear. Does it mean "Load all privilege info from database"?

Comment (tiancaiamao) [inline privilege/privileges/cache.go]: Yes, you're right.

File: plan/planbuilder.go [hunk 7]
```diff
@@ -69,8 +76,9 @@ type planBuilder struct {
 	inUpdateStmt bool
 	// colMapper stores the column that must be pre-resolved.
 	colMapper map[*ast.ColumnNameExpr]int
-
-	optFlag uint64
+	// Collect the visit information for privilege check.
+	visitInfo []visitInfo
```
Comment (hanfei1991) [inline plan/planbuilder.go]: Maybe `privlegeInfo` is a better name.

Comment (tiancaiamao) [inline plan/planbuilder.go]: I don't think so, `visitInfo` says nothing about privilege, why name it `privilegeInfo`.

**Comment by zimulala** [comment]
PTAL @hanfei1991 @coocood

**Pull request merged**